### PR TITLE
feat: NFT

### PR DESCRIPTION
**Pull request by green-jay**
## Description

Add the NFT pallet to provide basic functionality of non-fungible tokens for future more advanced usages.

## Related Issue

Fixes: #12

### CODE REVIEW

File: pallets/nft/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+		NFTTokenMinted(T::AccountId, T::ClassId, u32),
+		NFTTokenMintedLockToggled(T::AccountId, T::ClassId, T::TokenId, bool),
+		NFTTokenTransferred(T::AccountId, T::AccountId, T::ClassId, T::TokenId),
+		NFTTokenBurned(T::AccountId, T::ClassId, T::TokenId),
+		NFTTokenClassDestroyed(T::AccountId, T::ClassId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		ClassNotFound,
```
Comment (enthusiastmartin) [inline pallets/nft/src/lib.rs:147]: Add error docs strings for each error. They are used on the FE side.

File: pallets/nft/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+						   // fix clone
+				Self::deposit_event(Event::NFTTokenMintedLockToggled(
+					account.clone(),
+					token_id.0,
+					token_id.1,
+					token.data.locked,
+				));
+			}
+			Ok(())
+		})?;
```
Comment (enthusiastmartin) [inline pallets/nft/src/lib.rs:181]: I believe that you can just return from here directly. So the last Ok(()) won;t be necessary.

**Review by enthusiastmartin — COMMENTED** [review]
First quick review. I haven't checked all of it yet but first few comments might help.

I will continue again with the rest later.

File: node/src/command.rs [hunk 3]
```diff
@@ -15,12 +15,14 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#![allow(clippy::redundant_clone)]
```
[the remark this answers is not in the document]
Comment (green-jay) [reply node/src/command.rs]: it was this time...earlier it threw some errors, god knows why

File: pallets/nft/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			ensure!(quantity > Zero::zero(), Error::<T>::InvalidQuantity);
+			let class_info = orml_nft::Pallet::<T>::classes(class_id).ok_or(Error::<T>::ClassNotFound)?;
+			ensure!(sender == class_info.owner, Error::<T>::NoPermission);
+			let mut data = token_data;
+			data.locked = false;
+			for _ in 0..quantity {
+				orml_nft::Pallet::<T>::mint(&sender, class_id, metadata.clone(), data.clone())?;
+			}
```
[the remark this answers is not in the document]
Comment (green-jay) [reply pallets/nft/src/lib.rs:82]: if there is no performance difference, I would keep as it = better readability imho

File: pallets/nft/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+		}
+
+		#[pallet::weight(<T as Config>::WeightInfo::transfer())]
+		pub fn transfer(
+			origin: OriginFor<T>,
+			dest: <T::Lookup as StaticLookup>::Source,
+			token: (T::ClassId, T::TokenId),
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			let _class_info = orml_nft::Pallet::<T>::classes(token.0).ok_or(Error::<T>::ClassNotFound)?;
```
[the remark this answers is not in the document]
Comment (green-jay) [reply pallets/nft/src/lib.rs:94]: doesn't work

File: pallets/nft/src/lib.rs (hunk 2 above)
Comment (green-jay) [reply pallets/nft/src/lib.rs:186]: thought so too, but doesn't work

Comment (enthusiastmartin) [reply pallets/nft/src/lib.rs:186]:
```suggestion
		})
```

and removing the Ok(()) at the of the function  - that should be fine.

File: pallets/nft/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/nft/src/lib.rs:94]:
```suggestion
			ensure!(orml_nft::Pallet::<T>::classes(token.0).is_some(),Error::<T>::ClassNotFound);
```

ah sorry, i thought it was returning Result.  So in this way, it should be fine.

File: pallets/nft/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+		NFTTokenClassCreated(T::AccountId, T::ClassId),
+		NFTTokenMinted(T::AccountId, T::ClassId, u32),
+		NFTTokenMintedLockToggled(T::AccountId, T::ClassId, T::TokenId, bool),
+		NFTTokenTransferred(T::AccountId, T::AccountId, T::ClassId, T::TokenId),
+		NFTTokenBurned(T::AccountId, T::ClassId, T::TokenId),
+		NFTTokenClassDestroyed(T::AccountId, T::ClassId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
```
Comment (enthusiastmartin) [inline pallets/nft/src/lib.rs:143]: Tests for some error are missing.

**Review by enthusiastmartin — COMMENTED** [review]
few more comments. Still need to check benchmarking and tests.

File: pallets/nft/src/benchmarking.rs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	caller
+}
+
+benchmarks! {
+	create_class {
+		let caller = create_account::<T>("caller", 0);
+		let class_metadata = "just a token class".as_bytes().to_vec();
+		let class_data = "just some class data".as_bytes().to_vec();
+	}: _(RawOrigin::Signed(caller.clone()), class_metadata, class_data, T::CurrencyBalance::from(666_u128).into())
+	verify {
```
Comment (enthusiastmartin) [inline pallets/nft/src/benchmarking.rs:23]: I would add something to the verify sections of exchange benchmark too. It is helpful when you run tests.

File: pallets/nft/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+	pub type ClassBondUntil<T: Config> =
+		StorageDoubleMap<_, Twox64Concat, T::BlockNumber, Twox64Concat, ClassIdOf<T>, (), OptionQuery>;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + orml_nft::Config<ClassData = ClassData, TokenData = TokenData> {
+		type Currency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type WeightInfo: WeightInfo;
+		// This type is needed to convert from Currency to Balance
+		type CurrencyBalance: From<Balance>
```
Comment (enthusiastmartin) [inline pallets/nft/src/lib.rs:71]: I wonder if  this is really needed here.  It is not used anywhere else other than benchmarks.

File: pallets/nft/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+				Ok(())
+			})
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_finalize(now: T::BlockNumber) {
+			let bond = T::ClassBondAmount::get();
+			Self::unlock_bond(now, bond);
```
Comment (enthusiastmartin) [inline pallets/nft/src/lib.rs:205]: This should be have some weight too, no ? So it has to be taken into account.

File: pallets/nft/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+		Ok(token_info.data.locked)
+	}
+
+	pub fn toggle_lock(account: &T::AccountId, token_id: (T::ClassId, T::TokenId)) -> DispatchResult {
+		let _class_info = orml_nft::Pallet::<T>::classes(token_id.0).ok_or(Error::<T>::ClassNotFound)?;
+		orml_nft::Tokens::<T>::mutate_exists(token_id.0, token_id.1, |token| -> DispatchResult {
+			if let Some(ref mut token) = token {
+				ensure!(*account == token.owner, Error::<T>::NotTokenOwner);
+				token.data.locked ^= true; // Toggle
+						   // fix clone
```
Comment (enthusiastmartin) [inline pallets/nft/src/lib.rs:265]: clone is misbehaving ? ;-)

**Review by enthusiastmartin — COMMENTED** [review]
Only few comments or question. But looks ok to me.

File: node/src/service.rs [hunk 11]
```diff
@@ -37,7 +35,7 @@ pub type Block = sp_runtime::generic::Block<Header, sp_runtime::OpaqueExtrinsic>
 type Hash = sp_core::H256;
 
 pub fn new_partial(
-	config: &Configuration
+	config: &Configuration,
```
[the remark this answers is not in the document]
Comment (green-jay) [reply node/src/service.rs:38]: author cargo fmt :)

File: pallets/nft/src/lib.rs (hunk 8 above)
Comment (green-jay) [reply pallets/nft/src/lib.rs:71]: wondering the same thing. would you know the magic formula to use currency trait methods and pass balances to it?

File: pallets/nft/src/lib.rs (hunk 10 above)
Comment (green-jay) [reply pallets/nft/src/lib.rs:265]: removed

File: pallets/nft/src/lib.rs (hunk 8 above)
Comment (enthusiastmartin) [reply pallets/nft/src/lib.rs:71]: I'll have a look.

Comment (enthusiastmartin) [reply pallets/nft/src/lib.rs]: resolved #64

File: pallets/nft/src/lib.rs (hunk 9 above)
Comment (green-jay) [reply pallets/nft/src/lib.rs:209]: added to create_class weight calculation

File: runtime/src/lib.rs [hunk 12]
```diff
@@ -173,6 +175,7 @@ impl Filter<Call> for BaseFilter {
 			| Call::LBP(_)
 			| Call::Utility(_)
 			| Call::Vesting(_)
+			| Call::Nft(_)
```
Comment (jak-pan) [inline runtime/src/lib.rs:178]: Caps pls?

File: runtime/src/lib.rs [hunk 13]
```diff
@@ -676,15 +698,17 @@ construct_runtime!(
...
+		OrmlNft: orml_nft::{Pallet, Storage, Config<T>},
+		Tokens: orml_tokens::{Pallet, Storage, Call, Event<T>, Config<T>},
 
 		// Basilisk related modules
 		AssetRegistry: pallet_asset_registry::{Pallet, Call, Storage, Config<T>},
 		XYK: pallet_xyk::{Pallet, Call, Storage, Event<T>},
 		Exchange: pallet_exchange::{Pallet, Call, Storage, Event<T>},
 		LBP: pallet_lbp::{Pallet, Call, Storage, Event<T>},
 		MultiTransactionPayment: pallet_transaction_multi_payment::{Pallet, Call, Config<T>, Storage, Event<T>},
+		Nft: pallet_nft::{Pallet, Call, Event<T>, Storage},
```
Comment (jak-pan) [inline runtime/src/lib.rs:711]: Caps pls

File: runtime/src/lib.rs [hunk 14]
```diff
@@ -724,15 +746,17 @@ construct_runtime!(
 		AuraExt: cumulus_pallet_aura_ext::{Pallet, Config},
 
 		// ORML related modules
-		Tokens: orml_tokens::{Pallet, Storage, Call, Event<T>, Config<T>},
 		Currencies: orml_currencies::{Pallet, Call, Event<T>},
+		OrmlNft: orml_nft::{Pallet, Storage, Config<T>},
```
Comment (jak-pan) [inline runtime/src/lib.rs:750]:
```suggestion
		OrmlNFT: orml_nft::{Pallet, Storage, Config<T>},
```

File: node/src/chain_spec.rs [hunk 15]
```diff
@@ -363,6 +363,9 @@ fn parachain_genesis(
 			members: tech_committee_members,
 			phantom: Default::default(),
 		},
+		orml_nft: OrmlNftConfig {
```
Comment (jak-pan) [inline node/src/chain_spec.rs:366]: Ok we should set a guideline about naming abbreviations. I can see we're doing Id.

**Review by jak-pan — APPROVED** [review]

**Pull request merged**